ucb-rit/coldfront
# Enhance tracking of exceptions; correct possible issues with SavioProjectRequestWizard

**Fixes #366**

**Changes**
- Added a middleware `ExceptionMiddleware` that logs uncaught exceptions raised by views.
    - Because it was added last (or near last), it should be the first (or near first) middleware to run.
    - Many entries in the production log are: `ERROR log.log_response: Internal Server Error: /project/savio-project-request/`, but unexpectedly have no exception details; this middleware should include them.
- Integrated Sentry into the application and Ansible configuration.
    - Uncaught exceptions and log messages at or above the `ERROR` level are sent to Sentry.
        - Some loggers are ignored, since at most 5K events can be sent in the free tier per month.
            - `coldfront.api` includes some noisy messages.
            - `coldfront.core.utils.middleware` deals with exceptions that are already being sent, so the logged errors would be duplicates.
    - One percent of transactions (`traces_sample_rate`) are sent to Sentry.
        - At most 10k transactions can be sent in the free tier per month (and sending them induces load on the server).
        - Assuming 2.4 million jobs over 12 months, there are 200k jobs per month.
        - Assuming 3 transactions per job, there are 600k transactions per month.
        - 1% of 600k is 6k < 10k.
    - For now, the free Developer plan will be used for production, accessible to only one user, but this can be upgraded in the future if needed.
- Resolved potential issues with the view for submitting a request for a new Savio project.
    - Corrected the method signatures of some methods in classes that inherit from `SessionWizardView`.
    - Added a smoke test to ensure that submitting a request with valid inputs succeeds.
    - Increased the length of a user's session from 15 minutes to 1 hour.
        - One Internal Server Error for this view included `AttributeError: 'AnonymousUser' object has no attribute 'userprofile'`, suggesting that the user's session timed out before the form was successfully submitted.
- Set the `SESSION_COOKIE_SECURE` setting to `True` when SSL is enabled.

**How to Test**
- Review the code changes and add comments as needed.
- Ensure that the test suite passes.
    - At minimum, ensure that `coldfront.core.project.tests.test_views.test_savio_project_request_wizard` and `coldfront.core.project.tests.test_views.test_allocation_renewal_request_view` pass.
- Manual Testing
    - Include the new `sentry_dsn` variable from `main.copyme` in `main.yml`.
    - Re-run the Ansible playbook.
        - This will install `sentry-sdk`.
        - In `dev_settings.py`, ensure that `SESSION_COOKIE_SECURE` is set to `False` and that `SENTRY_DSN` is set to the empty string.
        - Ensure that the playbook runs successfully and that no obvious errors are raised on the website.
    - Raise an exception in a view (i.e., in the first line of the `done` method of `SessionWizardView`).
        - After restarting Apache, trigger the view.
        - Ensure that the uncaught exception was raised, _and_ that it was logged to `/var/log/user_portals/cf_mybrc_portal.log`, under `middleware.process_exception`.
